app/controllers/concerns/hud_reports/artifact_aware_cells.rb

```diff
@@ -0,0 +1,135 @@
+        merge(::HudReports::ReportCell.for_table(table).for_cell(cell)).
+        merge(::HudReports::ReportInstance.where(id: @report.id))
+    end
+
+    # Load clients using S3 data.
+    # This will only return clients for reports whose data HAS been stored in S3.
+    def load_clients_from_s3(client_class, cell, table)
+      return client_class.none unless cell
+
+      # This should work if we get the cell nam (i.e. "B3") or if we get the ReportCell object
+      report_cell = if cell.is_a?(String)
+        @report.report_cells.find_by(cell_name: cell, question: table)
+      else
+        cell.question == table ? cell : @report.report_cells.find_by(cell_name: cell.cell_name, question: table)
+      end
+
+      return client_class.none unless report_cell
+
+      # Pull the csv file for the universe members from the universe_members_csv file for this question
+      # NOte: These files are sharded by report question/table
+      service = HudReports::S3ArtifactService.new(@report)
+      universe_csv = service.retrieve_universe_members(question: table)
+      return client_class.none unless universe_csv
+
+      # Filter to the specific cell
+      cell_members = universe_csv.select { |row| row['report_cell_id'].to_i == report_cell.id }
+      member_ids = cell_members.map { |row| row['universe_membership_id'].to_i }.compact.uniq
+
+      # Load the actual report client records from the report_clients_csv file
+      # This file contains the client_id, source_enrollment_id, data_source_id, project_id, organization_id, and user_id
+      # for each client in the report.
+      clients_csv = service.retrieve_report_clients
+      return client_class.none unless clients_csv
+
+      # Filter the client datafor the specific universe members and create temporary objects to build out the cell data
+      matching_clients = clients_csv.select { |row| member_ids.include?(row['id'].to_i) }
+      return create_temporary_cell_objects(matching_clients)
+    end
+
+    def create_temporary_cell_objects(csv_rows)
+      # Create temporary objects from CSV data and attach preloaded warehouse clients and associations
+      temporary_objects = []
+
+      # Build out data that will be used in the cell data. These items are being preloaded as are being
+      # used in calculations within a HUD report when displaying the cell data (e.g. PII calculations or
+      # pulling names from these objects for a more user friendly display)
+      client_ids = csv_rows.map { |r| r['client_id']&.to_i }.compact.uniq
+      source_enrollment_ids = csv_rows.map { |r| r['source_enrollment_id']&.to_i }.compact.uniq
+      data_source_ids = csv_rows.map { |r| r['data_source_id']&.to_i }.compact.uniq
+      project_ids = csv_rows.map { |r| r['project_id']&.to_i }.compact.uniq
+      organization_ids = csv_rows.map { |r| r['organization_id']&.to_i }.compact.uniq
+      user_ids = csv_rows.map { |r| r['user_id']&.to_i }.compact.uniq
```

Comment:
This is looping over the CSV rows 6 times.  I think it'll be more efficient to do something like:

```suggestion
      client_ids = Set.new
      source_enrollment_ids = Set.new
      data_source_ids = Set.new
      project_ids = Set.new
      organization_ids = Set.new
      user_ids = Set.new
      
      csv_rows.each do |r| 
        client_ids << r['client_id']&.to_i 
        source_enrollment_ids << r['source_enrollment_id']&.to_i
        data_source_ids << r['data_source_id']&.to_i
        project_ids << r['project_id']&.to_i
        organization_ids << r['organization_id']&.to_i
        user_ids << r['user_id']&.to_i
      end
```

Reply:
Included in [this commit](https://github.com/greenriver/hmis-warehouse/pull/5644/commits/189c4cdec097138b0dbd1723976918957ed2e9d8)